grails-app/services/com/unifina/service/ChallengeService.groovy

```diff
@@ -7,36 +7,42 @@ import com.unifina.security.Challenge
 class ChallengeService {
 	static final int TTL_SECONDS = 300
 	static final int CHALLENGE_LENGTH = 30
+	private static final String textPart1 = "This is a challenge created by Streamr for address "
+	private static final String textPart2 = " to prove private key ownership by signing this random data with it:\n\n"
 
 	KeyValueStoreService keyValueStoreService
 
-	Challenge createChallenge() {
-		String challengeText = "This is a challenge created by Streamr to prove private key ownership by signing this random data with it:\n\n"
-		Challenge ch = new Challenge(challengeText, CHALLENGE_LENGTH, TTL_SECONDS)
+	Challenge createChallenge(String address) {
+		Challenge ch = new Challenge(getChallengeText(address), CHALLENGE_LENGTH, TTL_SECONDS)
 		keyValueStoreService.setWithExpiration(ch.id, ch.challenge, TTL_SECONDS)
 		return ch
 	}
 
-	boolean verifyChallengeResponse(String challengeID, String challenge, String signature, String publicKey) {
-		try {
-			String address = verifyChallengeAndGetAddress(challengeID, challenge, signature)
-			boolean valid = address.toLowerCase() == publicKey.toLowerCase()
-			if (valid) {
-				keyValueStoreService.delete(challengeID)
-			}
-			return valid
-		} catch(ChallengeVerificationFailedException e) {
-			return false
+	void checkValidChallengeResponse(String challengeID, String challenge, String signature, String givenAddress) throws ChallengeVerificationFailedException {
+		String address = verifyChallengeAndGetAddress(challengeID, challenge, signature)
+		boolean valid = address.toLowerCase() == givenAddress.toLowerCase() && challengeTextContainsAddress(challenge, givenAddress)
```

Comment:
Yes it is needed to prove private key ownership. Otherwise the following scenario is possible:

- User generates random address `A` and makes a `POST` on `api/v1/login/challenge/A`.
- Backend generates challenge text `ch` that contains address `A` and send it back to user.
- User generates a random signature `S` and ECrecover `A'` from `ch` and `S`.
- User sends `ch`, `S` and `A'` as his own address to `api/v1/login/response`.
- Backend checks that the received challenge text `ch` is the same that the one stored. It is.
- Backend ECrecover an address `A''` from `ch` and `S` and checks that `A'' == A'`. They are equal so user gets logged in.

Basically we need to remember the address given in the first place. But now that I think about it, the user has to give his address twice. I could change the spec to drop the address in the `api/v1/login/response`, just ECrecover `A'` from `ch` and `S` and extract `A` from `ch`, then check `A == A'`.

Reply:
@mthambipillai Sorry, my question was unclear. What I meant was, should we have:

`challengeTextContainsAddress(challenge, givenAddress.toLowerCase())`

instead of what you currently have (without `toLowerCase`):

`challengeTextContainsAddress(challenge, givenAddress)`